blocks/bricks/recurrent.py

```diff
@@ -630,3 +631,241 @@ def apply(self, *args, **kwargs):
+        for t in ["states", "contexts", "outputs"]:
+            v = [s + '_' + str(d)
+                 for d, transition in enumerate(transitions)
+                 for s in getattr(transition.apply, t)]
+            setattr(self.apply, t, v)
+
+        # sum up all the arguments we exepct to see in a call to a transition
+        # apply method, anything else is a recursion control
+        self.special_args = set(sum(map(lambda t: getattr(self.apply, t),
+                                        ["sequences", "states", "contexts"]),
+                                    ["mask"]))
+
+    def normal_inputs(self, d):
+        transition = self.transitions[d]
+        return [name for name in transition.apply.sequences
+                if 'mask' not in name]
+
+    def _push_allocation_config(self):
+        # Configure the forks that connect the "states" element in the `states`
+        # of one layer to the elements in the `sequences` of the next layer,
+        # excluding "mask".
+        # This involves `get_dim` requests
+        # to the transitions. To make sure that it answers
+        # correctly we should finish its configuration first.
+        for transition in self.transitions:
+            transition.push_allocation_config()
+
+        for d, fork in enumerate(self.forks):
+            fork.input_dim = self.transitions[d].get_dim(self.states_name)
+            fork.output_dims = self.transitions[d+1].get_dims(
+                fork.output_names)
+        for d, skip in enumerate(self.skip_connections):
+            skip.input_dims = self.transitions[0].get_dims(
+                self.normal_inputs(0))
+            skip.output_dims = self.transitions[d+1].get_dims(
+                self.normal_inputs(d+1))
+
+    def do_apply(self, *args, **kwargs):
+        """Apply the stack of transitions.
+
+        This is the undecorated implementation of the apply method.
+        It is separated from the decorated apply method in order to allow
+        usage of different docrations (wrappers) to be used.
+
+
+        Parameters
+        ----------
+        iterate : bool
+            If ``True`` iteration is made. By default ``True``.
+        reverse : bool
+            If ``True``, the sequences are processed in backward
+            direction. ``False`` by default.
+        return_initial_states : bool
+            If ``True``, initial states are included in the returned
+            state tensors. ``False`` by default.
+        The names appearing in the sequences, states, contexts parameters
+            of the apply method of each of the transitions in
+            self.transitions. Each such name is suffixed with a layer
+            number.
+
+        Returns
+        -------
+        The outputs of all transitions.
+        All return values are of type :class:`~tensor.TensorVariable`.
+
+        """
+        results = []
+        last_states = None
+        layer0_kwargs = None
```

Comment:
You can already initialize `layer0_kwargs` here, there is no need to do it inside the loop